pints/tests/test_opt_hager_zhang.py

```diff
@@ -0,0 +1,476 @@
+        proposed = m.ask()
+        reply = r0.evaluateS1(proposed)
+        m.tell([reply])
+        self.assertEqual(m._HagerZhang__initialising(k=0, alpha_k0=None),
+                         0.0024999999999999996)
+
+        # testing for starting point at all X are 0.0
+        del(opt)
+        del(m)
+        opt = pints.OptimisationController(r, x0, method=method)
+        m = opt.optimiser()
+        m.set_search_direction(px)
+        # using first ask tell to set up problem
+        proposed = m.ask()
+        reply = r.evaluateS1(proposed)
+        m.tell([reply])
+        self.assertEqual(m._HagerZhang__initialising(k=0, alpha_k0=None), 1.0)
+
+        # testing k != 0 i.e a second line search in a
+        # gradient based optimiser
+        self.assertEqual(m._HagerZhang__initialising(k=1, alpha_k0=3.0), 6.0)
+
+    def test__very_close(self):
+        # Tests the very close method of this optimiser.
+        r, x, s, b, px = self.problem()
+        opt = pints.OptimisationController(r, x, method=method)
+        m = opt.optimiser()
+        self.assertTrue(m._HagerZhang__very_close(2.0, (2.0 + 1E-17)))
+        self.assertFalse(m._HagerZhang__very_close(2.0, (2.0 + 1E-14)))
+
+    def test__position_after_step(self):
+        # Tests the __position_after_step method of this optimiser.
+        r, x, s, b, px = self.problem()
+        opt = pints.OptimisationController(r, x, method=method)
+        m = opt.optimiser()
+        m.set_search_direction(px)
+        logic, position = m._HagerZhang__position_after_step(alpha=0.5)
+        self.assertTrue(logic)
+        self.assertEqual(np.all(position), np.all([0.00015, 0.00015]))
+
+    def test__obj_and_grad_unpack(self):
+        # Tests the __obj_and_grad_unpack method of this optimiser.
+        r, x, s, b, px = self.problem()
+        opt = pints.OptimisationController(r, x, method=method)
+        m = opt.optimiser()
+        reply = r.evaluateS1(x)
+        func = reply[0]
+        grad = reply[1]
+        grad = np.matmul(np.transpose(grad), np.asarray(px))
+        m.set_search_direction(px)
+        self.assertEqual(np.all(
+                         m._HagerZhang__obj_and_grad_unpack(x=[reply])),
+                         np.all([func, grad]))
+
+    def test__bisect_or_secant(self):
+        # Tests the __bisect_or_secant method of this optimiser.
+        r, x, s, b, px = self.problem()
+        opt = pints.OptimisationController(r, x, method=method)
+        m = opt.optimiser()
+        m.set_search_direction(px)
+        # using first ask tell to set up problem
+        proposed = m.ask()
+        reply = r.evaluateS1(proposed)
+        m.tell([reply])
+        # grad = r.evaluateS1(x)[1]
+        # grad =np.matmul(np.transpose(grad), np.asarray([0.0001, 0.0001]))
+        # print('grad ',grad)
+        a = 0.5  # d = 0.75
+        b = 1.0
+
+        def temp(a=a, b=b, m=m):
+            generator = m._HagerZhang__bisect_or_secant(a=a, b=b)
```

Comment:
Not sure what the best way to test these is. _As a rule of thumb_ you don't test the private API. So we need to figure out if this is an exception or if there's something we're not seeing that would break this down into smaller testable blocks with public methods